### PR TITLE
feature: M107 Beaver engineering vehicle and vehicle trench placement changes

### PR DESCRIPTION
**Pull request by EmilyWinjum**
# ~~This PR is marked as a draft because it is: Ugly!~~ Oh yeah, it's all coming together
~~I wasn't sure how to clean this up in the right ways so please suffer with me as you look at a feature that's... functional!~~

Adds a new SWS vehicle (M107 Beaver, textures to come later) and enables vehicles with plows to select different trenches to build while their plows are down

### CODE REVIEW

File: addons/vehicleTrenches/functions/fnc_lowerPlow.sqf [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * [_vehicle, PLOW_NONE] call sws_vehicletrenches_fnc_lowerPlow
+ *
+ * Public: No
+ */
+
+params ["_vehicle", "_plowMode"];
+
+// Grab config for the vehicle's plow
+private _config = configFile >> "CfgVehicles" >> typeOf _vehicle >> "PlowConfig";
+if (!(isClass _config)) exitWith {
```
Comment (WitchFreya) [inline addons/vehicleTrenches/functions/fnc_lowerPlow.sqf:23]: either remove this check or make it `if !(isClass _config) exitWith {ERROR_1("Attempted to lowerPlow on invalid vehicle",_vehicle)};`

Comment (WitchFreya) [reply addons/vehicleTrenches/functions/fnc_lowerPlow.sqf:23]: Generally diag_log isn't to be used directly, prefer the cba macros for debug info

File: addons/vehicleTrenches/functions/fnc_lowerPlow.sqf [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		(_vehicle animationSourcePhase _animation == _plowLowered)
+	};
+},
+{	
+	// Apply plow mode on finish
+	params ["_vehicle", "_type", "_animation", "_plowLowered", "_plowMode"];
+	_vehicle setVariable [QGVAR(plowMode), _plowMode, true];
+	hint "Plow lowered.";
+	// Add frame event handler for the lowered plow
+	["addBuildHandler", [_vehicle]] call CBA_fnc_globalEvent;
```
Comment (WitchFreya) [inline addons/vehicleTrenches/functions/fnc_lowerPlow.sqf:55]: is addBuildHandler something in CBA or ACE Or something? if not, then use QGVAR for the name of the event.

File: addons/vehicles/CfgVehicles/Beaver.inc [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    };
+    hiddenSelectionsTextures[] = {
+        QPATHTOF(data\beaver\chassis.paa),
+        QPATHTOF(data\beaver\body.paa),
+        QPATHTOF(data\beaver\turret.paa),
+        QPATHTOF(data\beaver\crane.paa),
+		GREEN_NET_TEXTURE
+    };
+
+	class TextureSources : TextureSources{
```
Comment (WitchFreya) [inline addons/vehicles/CfgVehicles/Beaver.inc:26]:
```suggestion
	class TextureSources : TextureSources
  {
```

**Review by WitchFreya — CHANGES_REQUESTED** [review]
looked over, needs some fixes but good work over all; you can also remove `TrenchOffsets.inc` i believe.

File: addons/patches/gradtrenches/CfgFunctions.inc [hunk 4]
```diff
@@ -0,0 +1,17 @@
+class CfgFunctions
+{
+    class grad_trenches_functions
```
[the remark this answers is not in the document]
Comment (EmilyWinjum) [reply addons/patches/gradtrenches/CfgFunctions.inc:3]: I'm not sure what spacing needs to be cleaned here, what are you referring to specifically?

[the remark this answers is not in the document]
Comment (WitchFreya) [reply addons/patches/gradtrenches/CfgFunctions.inc:3]: ... i think github's previewer botched or something, disregard

File: addons/vehicleTrenches/functions/fnc_addBuildHandler.sqf [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * Example:
+ * [vehicle ACE_player] call sws_vehicletrenches_fnc_addBuildHandler
+ *
+ * Public: No
+ */
+
+params ["_vehicle"];
+
+private _config = configFile >> "CfgVehicles" >> typeOf _vehicle >> "PlowConfig";
```
Comment (WitchFreya) [inline addons/vehicleTrenches/functions/fnc_addBuildHandler.sqf:21]:
```suggestion
private _config = configOf _vehicle >> "PlowConfig";
```

File: addons/vehicleTrenches/functions/fnc_addBuildHandler.sqf [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+private _distanceToTrench = getNumber (_config >> "distanceToTrench");
```
Comment (WitchFreya) [inline addons/vehicleTrenches/functions/fnc_addBuildHandler.sqf:22]:
```suggestion
private _distanceToTrench = (_config >> "distanceToTrench") call BIS_fnc_getCfgData;
```

File: addons/vehicleTrenches/functions/fnc_continueTrench.sqf [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+params ["_vehicle", "_digDiff", "_flatOffsets", "_isTiltAboveMax", "_initMode", "_handle"];
+
+private _trench = _vehicle getVariable [QGVAR(trenchDigged), nil];
+private _actualProgress = _trench getVariable ["ace_trenches_progress", 0];
+
+// check for vehicle aliveness
+if (!(alive _vehicle)) exitWith {
+	[_handle] call CBA_fnc_removePerFrameHandler;
+	[_vehicle, _trench] call FUNC(detatchTrench);
```
Comment (WitchFreya) [inline addons/vehicleTrenches/functions/fnc_continueTrench.sqf:32]: spelling—`detachTrench`, not `detatch`.

File: addons/vehicleTrenches/XEH_postInit.sqf [hunk 8]
```diff
@@ -0,0 +1,5 @@
+#include "script_component.hpp"
+
+private _id = [QGVAR(addBuildHandler), { _this call FUNC(addBuildHandler); }] call CBA_fnc_addEventHandler;
+
+TRACE_1(Registered 'addBuildHandler',_id);
```
Comment (WitchFreya) [inline addons/vehicleTrenches/XEH_postInit.sqf:5]: ... does this work as-is? I'm shocked this didn't throw an error when building or something.
```suggestion
TRACE_1("Registered addBuildHandler",_id);
```

**Review by WitchFreya — CHANGES_REQUESTED** [review]
few more things, mostly good.

File: addons/vehicleTrenches/XEH_postInit.sqf (hunk 8 above)
Comment (EmilyWinjum) [reply addons/vehicleTrenches/XEH_postInit.sqf:5]: ... yeah I'm amazed that's actually been working at all lmao

**Comment by EmilyWinjum** [comment]
That last commit title felt like "Final.final.final.asdf1.mp4.final"

**Review by WitchFreya — APPROVED** [review]

**Pull request merged**